contracts/protocols/mainnet/morpho/aave_v3/helpers.sol

```diff
@@ -0,0 +1,320 @@
+            marketData_.aaveData.liquidityRate,
+            ,
+            ,
+            ,
+            ,
+            ,
+
+        ) = protocolData.getReserveData(asset);
+        return marketData_;
+    }
+
+    function getAaveHelperData(MarketDetail memory marketData_, address poolTokenAddress_)
+        internal
+        view
+        returns (MarketDetail memory)
+    {
+        marketData_.aaveData.totalSupplies = IAToken(poolTokenAddress_).totalSupply();
+        return marketData_;
+    }
+
+    function getAaveMarketData(
+        MarketDetail memory marketData_,
+        address underlying,
+        uint256 priceInEth,
+        uint256 priceInUsd
+    ) internal view returns (MarketDetail memory) {
+        marketData_.market = morphoGetter.market(underlying);
+        // marketData_.config.poolTokenAddress = marketData_.market.aToken;
+        marketData_.config.tokenPriceInEth = priceInEth;
+        marketData_.config.tokenPriceInUsd = priceInUsd;
+
+        marketData_ = getLiquidatyData(marketData_, marketData_.market.aToken);
+        marketData_ = getAaveHelperData(marketData_, marketData_.market.aToken);
+
+        return marketData_;
+    }
+
+    function getMarketData(
+        address underlying,
+        uint256 priceInEth,
+        uint256 priceInUsd
+    ) internal view returns (MarketDetail memory marketData_) {
+        marketData_ = getAaveMarketData(marketData_, underlying, priceInEth, priceInUsd);
+
+        //////////////////////////////////////////////////////
+        // (
+        //     marketData_.avgSupplyRatePerYear,
+        //     marketData_.avgBorrowRatePerYear,
+        //     marketData_.totalP2PSupply,
+        //     marketData_.totalP2PBorrows,
+        //     marketData_.totalPoolSupply,
+        //     marketData_.totalPoolBorrows
+        // ) = morphoGetter.getMainMarketData(underlying);
+
+        // (
+        //     marketData_.p2pSupplyRate,
+        //     marketData_.p2pBorrowRate,
+        //     marketData_.poolSupplyRate,
+        //     marketData_.poolBorrowRate
+        // ) = morphoGetter.getRatesPerYear(underlying);
+
+        // (
+        //     marketData_.p2pSupplyIndex,
+        //     marketData_.p2pBorrowIndex,
+        //     marketData_.poolSupplyIndex,
+        //     marketData_.poolBorrowIndex,
+        //     marketData_.lastUpdateTimestamp,
+        //     marketData_.p2pSupplyDelta,
+        //     marketData_.p2pBorrowDelta
+        // ) = morphoGetter.getAdvancedMarketData(underlying);
+        ///////////////////////////////////////////////////////
+    }
+
+    function getUserMarketData(
+        address user,
+        address underlying,
+        uint256 priceInEth,
+        uint256 priceInUsd
+    ) internal view returns (UserMarketData memory userMarketData_) {
+        userMarketData_.marketData = getMarketData(underlying, priceInEth, priceInUsd);
+
+        userMarketData_.p2pBorrows = morphoGetter.scaledP2PBorrowBalance(underlying, user);
```

Comment:
We need to convert the scaled values.
If you see, they are using `_getUserSupplyBalanceFromIndexes` inside `supplyBalance`. You can take the calculation logics from here.

<img width="966" alt="image" src="https://user-images.githubusercontent.com/47134275/223137286-e2079fae-8cd4-4e73-adf5-e80ac3ce482b.png">

Reply:
To get a logic of rayMulDown we should access here.
https://github.com/morpho-labs/morpho-data-structures-private/tree/06a9fb3fdf0d089977d99e5cd62774fb79f603a7
But I can't do it now.